Fix more HLR mask issues

jenshannoschwalm opened the pull request:
Still issues reported in #12913

The mask visuals were defined via 2 gui variables, they are now integrated into `dt_highlights_mask_t` for easier handling.

- 2 bugs in a--blown-segments were fixed, could generate false output data also in standard processing, definitely in the masks mode.
- strenght effect mask was not always safe for correctness
- make sure we don't write NaNs to output.

The pr should be concidered as a bugfix so far.

Not sure how to understand the issues seen in the video from @MStraeten

Martin could you check if this pr fixes your problems? (Under certain situations i was also able to see wrong output, not any more)

MStraeten commented:
unfortunately issue remains.
further investigations: i tried with several images and theres one commonality for images i can reproduce the scenario: they are taken with the same camera, canon 7D. i haven't found a scenario with my older and also newer models (just plain raw used).  
Issue even reproducible where i don't expect any clipping: [20100317-IMG_1234.CR2](https://mega.nz/file/6KwTGTrC#BgoAkTJO3MBLdkc1GYW-OKeBLiPiEi8Q4kEz2rwic8E)

while checking some images in the the cr3 raw samples package used when cr3 support was included first i found further used cameras where i can reproduce this behaviour: R, R5 with CRAW or sensor crop -> so will do some further investigations with that sample set.

jenshannoschwalm commented:
Will try to find out what's going on. I tested your image here and can't find anything unusual.

jenshannoschwalm commented:
There has been another issue fixed although i don't think that's your issue here as that problem would likely be possible on all images & platforms.

I sort of tested a hi-res monitor (switching in gnome to 200%) no issue here. Also checked on all demosaicers and scaling interpolators.

@parafin as you are also on mac, can you reproduce the issue?

MStraeten commented:
it's a bit better - i get a plausible mask view quite more often, sometimes it needs to switch between masks for combine and candidating. 
If the initial mask is fine then turning masking on/off several times gives different results (even a plausible mask again) 
Did a test with a windows notebook attached to the WQHD display - no change to reproduce the behaviour...

further investigations i'll do now:
clean darktable config, clean caches, new build with type Debug ...

MStraeten commented:
after cleaning cache and darktablerc the issue is no longer reproducible - when restoring the old darktablerc it's back again.
So i investigated some time to "bisect" in darktablerc:
it breaks: if plugins/darkroom/histogram/mode=waveform or parade

TurboGit commented:
> So i investigated some time to "bisect" in darktablerc:
> it breaks: if plugins/darkroom/histogram/mode=waveform or parade

Whaooo!!! Thanks for tracking this down.

parafin commented:
> @parafin as you are also on mac, can you reproduce the issue?

Is it still needed? If so, could you be specific which commit to check? If I find time, I may try to reproduce it.

jenshannoschwalm commented:
Whoau from me too. BTW I found another issue, thanks for pushing me so hard!

jenshannoschwalm commented:
I could find two problems in the code i fixed, the one relevant for the reported issue was a maybe-missing invalidate cache, refactored that code making sure the cache in cleared whenever a masking mode changed.

About the problem possibly identified by @MStraeten, maybe that could be related to the `DT_DEV_PIXELPIPE_DISPLAY_PASSTHRU` mode. Martin, could you please check
1. this pr
2. this pr after disabling the passthru mode (Lines 296 and 1993 in highlights.c)

MStraeten commented:
1. now the to be behaviour is ways better - mostly the to be behaviour is active on click, but after a few togglings of mask mode the strange behaviour can occur again, but also get rid again. But the behaviour seems to be dependent on the screen resolution: integrated macbook display quite good, maximized on external wqhd screen keeps being weird, half size on externals creen quite good. Need to do mor einvestigations in the afternoon.
2. disabling both lines is worse

jenshannoschwalm commented:
What do you mean by worse? Sure the output looks different but the reported artefacts?

MStraeten commented:
while 1 seems to have the artifacts as an exception with 2 the 'to be' scenario is an exception. Most time the artifacts occurs

MStraeten commented:
on macbookpro Retina:

https://user-images.githubusercontent.com/39386816/202858930-87832a4e-4ca6-41d9-99a4-5cec777c3666.mp4

on WQHD Screen partial

https://user-images.githubusercontent.com/39386816/202858944-8afac24d-1c02-44a2-9100-738bc8d93c8c.mp4

on WQHD Screen Fulscreen

https://user-images.githubusercontent.com/39386816/202858964-95c6df15-6608-4ee7-a159-1ccf005346ca.mp4

MStraeten commented:
my sample:
Full Screen:

https://user-images.githubusercontent.com/39386816/202859714-a93c0c21-175e-4f60-9bc3-a69d30c4dd97.mp4

partial screen:


https://user-images.githubusercontent.com/39386816/202859739-26d1628f-9e4e-4f9c-9290-6ed56eaa37d0.mp4

even if there's a plausible result this changes when toggling  masks on / off - but just in combination with waveform&parade

jenshannoschwalm commented:
@Martin i looked at all the videos - i have never seen anythink like **that** - making it impossible to test right now - and i confess i have currently no concrete idea what could be wrong. Just to make sure

1. You can con reproduce on osx?
2. you can reproduce on windows?
3. it only happens on screens with ppd > 1 (hires )
4. it is only visible with special modes in the histogramm?


You video shows different output just by toggling one of the masking buttons on ->off->on again. Either a module later in the pipeline gets wrong iop cached data or ... i simply dont know. The border issue - the bright red border around in the darkroom window - no idea how that can be.

MStraeten commented:
1 - i can reproduce it on osx 
2 - i can't reproduce it on windows (but thats a different system)
3 - it's also reproducible if i set FullHD resolution in macOS system settings for the WQHD screen
4 - it's just visible with histogram waveform and parade mode  

so at least there's a workaround for it - avoid those histogram modes

here two debug message files: 
quite_good.txt: from toggling mask to proper display,
weird.txt: from toggling mask to weird display
but they seems to be very similar
[quite_good.txt](https://github.com/darktable-org/darktable/files/10047892/quite_good.txt)
[weird.txt](https://github.com/darktable-org/darktable/files/10047893/weird.txt)

jenshannoschwalm commented:
For me your logs strongly hint to a pixelpipe cacheline problem - we seem to get a hit for a correct one but that's not true. 

@TurboGit there have been some definitive bugs fixed in this pr so i would propose to merge this as is.

I would fix - or at least try to do so - remaining issues in a seperate pr.

Also @MStraeten - you might want to check my iop cache ideas using this

[pixelpipe_cache.c.txt](https://github.com/darktable-org/darktable/files/10048942/pixelpipe_cache.c.txt)


1. the cacheline hit test checks for equal size instead of "enough" which is more correct
2. the basic hash also includes the passthru modes in mask as these directly affect output

TurboGit reviewed (approved):
Thanks for the hard work!

MStraeten commented:
checked with pixelpipe_cache.c.txt - keeps being weird.
here the log from darktable -d dev matching to the screenrecording
[debug_log.txt](https://github.com/darktable-org/darktable/files/10049254/debug_log.txt)
blank lines indicates the switch to vectorscope mode


https://user-images.githubusercontent.com/39386816/202894109-b68fef01-b8c5-4b81-beaf-74c880949c27.mp4


will keep an eye on #12945

jenshannoschwalm commented:
@MStraeten i think i might have found "something" completely overseen so far.

Could you remove L. 2159 in highlights.c ? and check?
```
  if(piece->pipe->mask_display & DT_DEV_PIXELPIPE_DISPLAY_MASK) dt_iop_alpha_copy(ivoid, ovoid, roi_out->width, roi_out->height);
```

ATM I think this completely wrong as all mask visualizers in this module don't work with an alpha copy.

MStraeten commented:
unfortunately it keeps being weird

jenshannoschwalm commented:
@MStraeten some - again very old - bugs in iop cache have been fixed.

Now
1. Please disable all modules after demosaic in the pipeline to make sure no other module interferes with not handling the passthru mode correctly.
2. use `-d roi -d dev -d verbose` and check log.

[log.txt](https://github.com/darktable-org/darktable/files/10050343/log.txt)

The interesting parts look like those starting in L.28 when toggling one of the mask visualizers in hlr.

The preview pipe should likely have a hit, if not it should show a `[proces ...] highlights` line.
On the contrary, the full pipe should **never** (by design right now as we invalidate all cachelines) have a hit when toggling the button and should **always** have a `[process ...] highlights` line.

If you can confirm this we are almost completely sure the issue is not in the hlr module.

MStraeten commented:
ok, trying to log gave a further aspect:
if darktable app loses focus before toggling mask on (e.g. during moving over to the terminal to enter some space into the log, or just activating another window), then toggling mask on after switching back to darktable results in an good result. If masks is toggled of and on again ther result is weird.

1. toggle on mask after giving darktable window focus --> ok
2. toggle off mask and switch over to terminal to enter some empty lines
3. toggle on mask after switching back to darktable window --> ok
4. toggle of mask  and don't leave darktable
5. toggle on mask --> weird
here the log (steps as inline comments)
[log3.txt](https://github.com/darktable-org/darktable/files/10050596/log3.txt)

jenshannoschwalm commented:
What image are you testing? Surprised by the number of segments, would like to check that.

About dt app loosing focus and issue gone - - absolutely crazy for me. Could be anything.

MStraeten commented:
It’s the image I sent to you some time ago (29.08.) before segmentation based highlight recovery was merged. Can‘t share it here …

jenshannoschwalm commented:
OK fine.
